### PR TITLE
Fix the fail start of tiflash because of the zero division when getting cpu number

### LINKED ISSUE #9212: Failed to start TiFlash when some CPUs are offline
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

In a machine with 72 logical cpu cores, and let 64 of them off-line.

6qzOnG0kEG

### 2. What did you expect to see? (Required)

Start TiFlash successfully.

### 3. What did you see instead (Required)

```
[2024/07/10 14:55:07.141 +08:00] [ERROR] [BaseDaemon.cpp:495] ["Integer divide by zero."] [source=BaseDaemon] [thread_id=2][2024/07/10 14:55:10.890 +08:00] [ERROR] [BaseDaemon.cpp:560] ["\n  0x563cb525a5ce\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+123495886]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:211\n  0x7fda5033b630\t<unknown symbol> [libpthread.so.0+63024]\n  0x563cafc6e83e\tcomputeAndSetNumberOfPhysicalCPUCores(unsigned short, unsigned short) [tiflash+33400894]\n                \tdbms/src/Common/getNumberOfCPUCores.cpp:66\n  0x563cafc39345\tDB::Server::main(std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&) [tiflash+33182533]\n                \tdbms/src/Server/Server.cpp:1077\n  0x563cb6fc7776\tPoco::Util::Application::run() [tiflash+154351478]\n                \tcontrib/poco/Util/src/Application.cpp:335\n  0x563cafc3306f\tDB::Server::run() [tiflash+33157231]\n                \tcontrib/poco/Util/src/ServerApplication.cpp:94\n  0x563cafc4c562\tmainEntryClickHouseServer(int, char**) [tiflash+33260898]\n                \tdbms/src/Server/Server.cpp:1855\n  0x563cafb82010\tmain [tiflash+32432144]\n                \tdbms/src/Server/main.cpp:172\n  0x7fda4faaa555\t__libc_start_main [libc.so.6+140629]"] [source=BaseDaemon] [thread_id=2]
[2024/07/10 14:55:34.790 +08:00] [INFO] [BaseDaemon.cpp:1175] ["Welcome to TiFlash"] [thread_id=1][2024/07/10 14:55:34.790 +08:00] [INFO] [BaseDaemon.cpp:1178] ["TiFlash build info: TiFlash\nRelease Version: v8.2.0-alpha-90-g914aaf71a\nEdition:         Community\nGit Commit Hash: 914aaf71ae9ffae2b9162dbf77c451bf5872ba2f\nGit Branch:      HEAD\nUTC Build Time:  2024-07-10 02:11:01\nEnable Features: jemalloc sm4(GmSSL) avx2 avx512 unwind thinlto\nProfile:         RELWITHDEBINFO\nCompiler:        clang++ 17.0.6\n"] [thread_id=1]
```

### 4. What is your TiFlash version? (Required)

master

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #9212

Problem Summary:

### What is changed and how it works?

```commit-message

```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the that fail to start tiflash because of the zero division when getting cpu number
```

### CODE REVIEW

File: dbms/src/Common/getNumberOfCPUCores.cpp [hunk 1]
```diff
@@ -62,8 +64,11 @@ void computeAndSetNumberOfPhysicalCPUCores(
...
     // - `number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores)` means how many physical cpu cores the tiflash process could use. (Actually, it's needless to get physical cpu cores in virtual environment, but we must ensure the behavior `1` is not broken)
     auto hardware_logical_cpu_cores = std::thread::hardware_concurrency();
-    UInt16 physical_cpu_cores
-        = number_of_logical_cpu_cores_ / (hardware_logical_cpu_cores / number_of_hardware_physical_cores);
+
+    UInt16 cpu_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
+    cpu_num_per_physical_core = std::max<UInt16>(1, cpu_num_per_physical_core);
+
+    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / cpu_num_per_physical_core;
     CPUCores::number_of_physical_cpu_cores = physical_cpu_cores > 0 ? physical_cpu_cores : 1;
```
Comment (Lloyd-Pottiger) [inline dbms/src/Common/getNumberOfCPUCores.cpp:72]:
```suggestion
    UInt16 thread_num_per_physical_core = hardware_logical_cpu_cores / number_of_hardware_physical_cores;
    thread_num_per_physical_core = std::max<UInt16>(2, cpu_num_per_physical_core); // ? maybe 2 is better?

    UInt16 physical_cpu_cores = number_of_logical_cpu_cores_ / thread_num_per_physical_core;
    CPUCores::number_of_physical_cpu_cores = std::max<UInt16>(1, physical_cpu_cores);
```

Comment (xzhangxian1008) [reply dbms/src/Common/getNumberOfCPUCores.cpp:72]: > thread_num_per_physical_core

Done. I think 1 is better as there may still be some cpu without hyper-threading.

**Review by Lloyd-Pottiger — APPROVED** [review]

File: dbms/src/Common/getNumberOfCPUCores.cpp (hunk 1 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Common/getNumberOfCPUCores.cpp]: [Hyper-threading](https://en.wikipedia.org/wiki/Hyper-threading) was introduced in 2002.

File: dbms/src/Common/getNumberOfCPUCores.cpp [hunk 2]
```diff
@@ -46,6 +46,8 @@ void computeAndSetNumberOfPhysicalCPUCores(
     UInt16 number_of_logical_cpu_cores_,
     UInt16 number_of_hardware_physical_cores)
 {
+    number_of_hardware_physical_cores = std::max<UInt16>(1, number_of_hardware_physical_cores);
```
Comment (yibin87) [inline dbms/src/Common/getNumberOfCPUCores.cpp:49]: Could you please describe what're the acutual parameter values when the issue happened?

Comment (xzhangxian1008) [reply dbms/src/Common/getNumberOfCPUCores.cpp:49]: > Could you please describe what're the acutual parameter values when the issue happened?

When the issue happens, hardware_logical_cpu_cores is lower than number_of_hardware_physical_cores which causes (hardware_logical_cpu_cores / number_of_hardware_physical_cores) is equal to 0.

**Pull request merged**